source/common/config/type_to_endpoint.cc

```diff
@@ -6,96 +6,170 @@
 
 // API_NO_BOOST_FILE
 
+#define API_VERSION envoy::config::core::v3::ApiVersion
+#define VERSIONED(v2, v3)                                                                          \
+  createServiceToVersionedMap(v2, API_VERSION::V2, {v2, v3}),                                      \
+      createServiceToVersionedMap(v3, API_VERSION::V3, {v2, v3})
+
 namespace Envoy {
 namespace Config {
 
 namespace {
 
-// service RPC method fully qualified names.
-struct Service {
-  std::string sotw_grpc_method_;
-  std::string delta_grpc_method_;
-  std::string rest_method_;
+struct VersionedMap {
+  envoy::config::core::v3::ApiVersion version_;
+  std::unordered_map<envoy::config::core::v3::ApiVersion, std::string> entries_;
+};
+
+// Map from resource type URL to versioned service RPC methods.
+using ServiceToVersionedMap = std::unordered_map<std::string, VersionedMap>;
+using TypeUrlVersionMap = std::unordered_map<std::string, envoy::config::core::v3::ApiVersion>;
+using VersionedMethodMap = std::unordered_map<envoy::config::core::v3::ApiVersion, std::string>;
+
+ServiceToVersionedMap
+createServiceToVersionedMap(absl::string_view service, envoy::config::core::v3::ApiVersion version,
+                            const std::array<std::string, 2>& versioned_services) {
+  return {{static_cast<std::string>(service),
+           VersionedMap{version,
```

Comment:
This is a bit confusing to the reader. Why is there a version, which then maps to more versions and services? Is one of these supported to be transport and one resource version? It's a bit unclear.

Reply:
This is the list of maps (as the input):

```js

[
    {
      "envoy.api.v2.RouteDiscoveryService": {
        "version_": "V2",
        "entries_": {
          "V2": "envoy.api.v2.RouteDiscoveryService",
          "V3": "envoy.service.route.v3.RouteDiscoveryService"
        }
      },
      "envoy.service.route.v3.RouteDiscoveryService": {
        "version_": "V3",
        "entries_": {
          "V2": "envoy.api.v2.RouteDiscoveryService",
          "V3": "envoy.service.route.v3.RouteDiscoveryService"
        }
      }
    },
    ...
]
```

Transformed into:

```js
{
  "type.googleapis.com/envoy.api.v2.RouteConfiguration": {
    "sotw_grpc_": {
      "type_url_versions_": {
        "type.googleapis.com/envoy.api.v2.RouteConfiguration": "V2",
        "type.googleapis.com/envoy.config.route.v3.RouteConfiguration": "V3"
      },
      "methods_": {
        "V2": "envoy.api.v2.RouteDiscoveryService.StreamRoutes",
        "V3": "envoy.service.route.v3.RouteDiscoveryService.StreamRoutes"      
      },
      ...
    }
  },
  "type.googleapis.com/envoy.config.route.v3.RouteConfiguration": {
    "sotw_grpc_": {
      "type_url_versions_": {
        "type.googleapis.com/envoy.api.v2.RouteConfiguration": "V2",
        "type.googleapis.com/envoy.config.route.v3.RouteConfiguration": "V3"
      },
      "methods_": {
        "V2": "envoy.api.v2.RouteDiscoveryService.StreamRoutes",
        "V3": "envoy.service.route.v3.RouteDiscoveryService.StreamRoutes"
      }
    },
    ...
  }
  ...
}


```

So it is convenient to get the effective `transport_api_version` when the value is`AUTO`: given a `resource_type_url` (e.g. `type.googleapis.com/envoy.api.v2.RouteConfiguration`) and `type_url_versions_` map (we carry this along for each entry).

```
effective_version = type_url_versions_[resource_type_url]
```

Reply:
It's probably fine as is, but I'd suggest adding in some type aliases for the keys that are `std::string`, so it's clear what is being keyed on, and also adding to the code comments some examples like above.

The other alternative is to compute all of this at runtime, instead of building all the maps, but it's unclear to me if that would be any simpler.